publish().autoConnect(0) behaving different with 3.3.0.M1

### Expected behavior
The test to succeed (as it does with reactor-core 3.2.6)

### Actual behavior
Test fails
```
Exception in thread "main" java.lang.AssertionError: expectation "expectNext(99)" failed (expected value: 99; actual value: 1)
```

### Steps to reproduce
```
        TestPublisher<Integer> publisher = TestPublisher.create();
        Flux<Integer> flux = publisher.flux().publish().autoConnect(0);
        publisher.next(1);
        publisher.next(2);
        publisher.next(3);

        StepVerifier.create(flux)
                    .expectSubscription()
                    .then(() -> publisher.next(99))
                    .expectNext(99)
                    .then(publisher::complete)
                    .verifyComplete();
```
### Reactor Core version
3.3.0.M1

### JVM version (e.g. `java -version`)
```
java version "1.8.0_191"
Java(TM) SE Runtime Environment (build 1.8.0_191-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.191-b12, mixed mode)
```

--- a/reactor-core/src/main/java/reactor/core/publisher/EmitterProcessor.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/EmitterProcessor.java
@@ -32,7 +32,6 @@ import reactor.core.Scannable;
 import reactor.util.annotation.Nullable;
 import reactor.util.concurrent.Queues;
 
-import static reactor.core.publisher.FluxPublish.PublishSubscriber.EMPTY;
 import static reactor.core.publisher.FluxPublish.PublishSubscriber.TERMINATED;
 
 /**
@@ -53,6 +52,9 @@ import static reactor.core.publisher.FluxPublish.PublishSubscriber.TERMINATED;
  * @author Stephane Maldini
  */
 public final class EmitterProcessor<T> extends FluxProcessor<T, T> {
+
+	@SuppressWarnings("rawtypes")
+	static final FluxPublish.PubSubInner[] EMPTY = new FluxPublish.PublishInner[0];
 
 	/**
 	 * Create a new {@link EmitterProcessor} using {@link Queues#SMALL_BUFFER_SIZE}
@@ -150,6 +152,8 @@ public final class EmitterProcessor<T> extends FluxProcessor<T, T> {
 		}
 		this.autoCancel = autoCancel;
 		this.prefetch = prefetch;
+		//doesn't use INIT/CANCELLED distinction, contrary to FluxPublish)
+		//see remove()
 		SUBSCRIBERS.lazySet(this, EMPTY);
 	}
 
@@ -535,6 +539,8 @@ public final class EmitterProcessor<T> extends FluxProcessor<T, T> {
 				System.arraycopy(a, j + 1, b, j, n - j - 1);
 			}
 			if (SUBSCRIBERS.compareAndSet(this, a, b)) {
+				//contrary to FluxPublish, there is a possibility of auto-cancel, which
+				//happens when the removed inner makes the subscribers array EMPTY
 				if (autoCancel && b == EMPTY && Operators.terminate(S, this)) {
 					if (WIP.getAndIncrement(this) != 0) {
 						return;

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java
@@ -335,7 +335,7 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 					return;
 				}
 				int n = a.length;
-				int j = 0;
+				int j = -1;
 				for (int i = 0; i < n; i++) {
 					if (a[i] == inner) {
 						j = i;
@@ -343,7 +343,8 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 					}
 				}
 
-				if (j < 0) { //TODO investigate condition always false
+				if (j < 0) {
+					//inner was not found
 					return;
 				}
 

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxPublish.java
@@ -179,8 +179,12 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 				AtomicIntegerFieldUpdater.newUpdater(PublishSubscriber.class,
 						"connected");
 
+		//notes: FluxPublish needs to distinguish INIT from CANCELLED in order to correctly
+		//drop values in case of an early connect() without any subscribers.
 		@SuppressWarnings("rawtypes")
-		static final PubSubInner[] EMPTY      = new PublishInner[0];
+		static final PubSubInner[] INIT       = new PublishInner[0];
+		@SuppressWarnings("rawtypes")
+		static final PubSubInner[] CANCELLED  = new PublishInner[0];
 		@SuppressWarnings("rawtypes")
 		static final PubSubInner[] TERMINATED = new PublishInner[0];
 
@@ -202,7 +206,7 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 		PublishSubscriber(int prefetch, FluxPublish<T> parent) {
 			this.prefetch = prefetch;
 			this.parent = parent;
-			SUBSCRIBERS.lazySet(this, EMPTY);
+			SUBSCRIBERS.lazySet(this, INIT);
 		}
 
 		boolean isTerminated(){
@@ -296,11 +300,12 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 				if (WIP.getAndIncrement(this) != 0) {
 					return;
 				}
-				disconnectAction(terminate());
+				disconnectAction();
 			}
 		}
 
-		void disconnectAction(PubSubInner<T>[] inners) {
+		void disconnectAction() {
+			PubSubInner<T>[] inners = SUBSCRIBERS.getAndSet(this, CANCELLED);
 			if (inners.length > 0) {
 				queue.clear();
 				CancellationException ex = new CancellationException("Disconnected");
@@ -331,7 +336,7 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 		public void remove(PubSubInner<T> inner) {
 			for (; ; ) {
 				PubSubInner<T>[] a = subscribers;
-				if (a == TERMINATED || a == EMPTY) {
+				if (a == TERMINATED || a == CANCELLED) {
 					return;
 				}
 				int n = a.length;
@@ -350,7 +355,7 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 
 				PubSubInner<?>[] b;
 				if (n == 1) {
-					b = EMPTY;
+					b = CANCELLED;
 				}
 				else {
 					b = new PubSubInner<?>[n - 1];
@@ -358,6 +363,8 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 					System.arraycopy(a, j + 1, b, j, n - j - 1);
 				}
 				if (SUBSCRIBERS.compareAndSet(this, a, b)) {
+					//we don't assume autoCancel semantics, which will rather depend from
+					//downstream operators like autoConnect vs refCount, so we don't disconnect here
 					return;
 				}
 			}
@@ -393,7 +400,7 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 
 				PubSubInner<T>[] a = subscribers;
 
-				if (a != EMPTY && !empty) {
+				if (a != CANCELLED && !empty) {
 					long maxRequested = Long.MAX_VALUE;
 
 					int len = a.length;
@@ -499,7 +506,7 @@ final class FluxPublish<T> extends ConnectableFlux<T> implements Scannable {
 
 		boolean checkTerminated(boolean d, boolean empty) {
 			if (s == Operators.cancelledSubscription()) {
-				disconnectAction(terminate());
+				disconnectAction();
 				return true;
 			}
 			if (d) {

--- a/reactor-core/src/test/java/reactor/core/publisher/EmitterProcessorTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/EmitterProcessorTest.java
@@ -907,4 +907,20 @@ public class EmitterProcessorTest {
 		            .expectComplete()
 		            .verify(Duration.ofSeconds(4));
 	}
+
+	@Test
+	public void removeUnknownInnerIgnored() {
+		EmitterProcessor<Integer> processor = EmitterProcessor.create();
+		EmitterProcessor.EmitterInner<Integer> inner = new EmitterProcessor.EmitterInner<>(null, processor);
+		EmitterProcessor.EmitterInner<Integer> notInner = new EmitterProcessor.EmitterInner<>(null, processor);
+
+		processor.add(inner);
+		assertThat(processor.subscribers).as("adding inner").hasSize(1);
+
+		processor.remove(notInner);
+		assertThat(processor.subscribers).as("post remove notInner").hasSize(1);
+
+		processor.remove(inner);
+		assertThat(processor.subscribers).as("post remove inner").isEmpty();
+	}
 }

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxPublishTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxPublishTest.java
@@ -31,6 +31,7 @@ import reactor.core.Scannable;
 import reactor.core.scheduler.Schedulers;
 import reactor.test.StepVerifier;
 import reactor.test.publisher.FluxOperatorTest;
+import reactor.test.publisher.TestPublisher;
 import reactor.test.subscriber.AssertSubscriber;
 import reactor.util.concurrent.Queues;
 
@@ -639,7 +640,7 @@ public class FluxPublishTest extends FluxOperatorTest<String, String> {
 	}
 
 	//see https://github.com/reactor/reactor-core/issues/1528
-	@Test
+	@Test(timeout = 4000)
 	public void syncFusionFromInfiniteStream() {
 		final ConnectableFlux<Integer> publish =
 				Flux.fromStream(Stream.iterate(0, i -> i + 1))
@@ -654,7 +655,7 @@ public class FluxPublishTest extends FluxOperatorTest<String, String> {
 	}
 
 	//see https://github.com/reactor/reactor-core/issues/1528
-	@Test
+	@Test(timeout = 4000)
 	public void syncFusionFromInfiniteStreamAndTake() {
 		final Flux<Integer> publish =
 				Flux.fromStream(Stream.iterate(0, i -> i + 1))
@@ -666,6 +667,42 @@ public class FluxPublishTest extends FluxOperatorTest<String, String> {
 		            .expectNextCount(10)
 		            .expectComplete()
 		            .verify(Duration.ofSeconds(4));
+	}
+
+	@Test
+	public void dataDroppedIfConnectImmediately() {
+		TestPublisher<Integer> publisher = TestPublisher.create();
+		ConnectableFlux<Integer> connectableFlux = publisher.flux().publish();
+
+		connectableFlux.connect();
+
+		publisher.next(1);
+		publisher.next(2);
+		publisher.next(3);
+
+		StepVerifier.create(connectableFlux)
+		            .expectSubscription()
+		            .then(() -> publisher.next(99))
+		            .expectNext(99)
+		            .then(publisher::complete)
+		            .verifyComplete();
+	}
+
+	@Test
+	public void dataDroppedIfAutoconnectZero() {
+		TestPublisher<Integer> publisher = TestPublisher.create();
+		Flux<Integer> flux = publisher.flux().publish().autoConnect(0);
+
+		publisher.next(1);
+		publisher.next(2);
+		publisher.next(3);
+
+		StepVerifier.create(flux)
+		            .expectSubscription()
+		            .then(() -> publisher.next(99))
+		            .expectNext(99)
+		            .then(publisher::complete)
+		            .verifyComplete();
 	}
 
 	@Test

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxPublishTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxPublishTest.java
@@ -667,4 +667,20 @@ public class FluxPublishTest extends FluxOperatorTest<String, String> {
 		            .expectComplete()
 		            .verify(Duration.ofSeconds(4));
 	}
+
+	@Test
+	public void removeUnknownInnerIgnored() {
+		FluxPublish.PublishSubscriber<Integer> subscriber = new FluxPublish.PublishSubscriber<>(1, null);
+		FluxPublish.PublishInner<Integer> inner = new FluxPublish.PublishInner<>(null);
+		FluxPublish.PublishInner<Integer> notInner = new FluxPublish.PublishInner<>(null);
+
+		subscriber.add(inner);
+		assertThat(subscriber.subscribers).as("adding inner").hasSize(1);
+
+		subscriber.remove(notInner);
+		assertThat(subscriber.subscribers).as("post remove notInner").hasSize(1);
+
+		subscriber.remove(inner);
+		assertThat(subscriber.subscribers).as("post remove inner").isEmpty();
+	}
 }

